File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        LinkedList<Character> stack = new LinkedList<>();
        int N = s.length();
        HashMap<Character, Character> left_par = new HashMap<>(){{put('(', ')'); put('{','}'); put('[',']');}};
        for(int i = 0; i < N; i++){
            char cur_char = s.charAt(i);
            if(stack.isEmpty() && left_par.containsKey(cur_char)){
                stack.add(cur_char);
            }else if(stack.isEmpty()){
                return false;
            }
            else{
                char last = stack.get(stack.size() - 1);
                if (left_par.containsKey(cur_char)){
                    stack.add(cur_char);
                }else{
                    if(left_par.get(last) != cur_char) return false;
                    stack.pollLast();

                }
            }
            }
        if(stack.isEmpty())return true;
        return false;
    }
        
    }

File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */

var map = new Map()
map.set('(', ')')
map.set('[',']')
map.set('{', '}')


var isValid = function(s) {
    var stack = []
    var n = s.length
    
    for (var i = 0; i < n; i++) {
        var char = s.charAt(i)
       // console.log(stack)
        if (map.has(char)) {
            stack.push(char)
        }else{
            if (stack.length == 0)
                return false
            var popped = stack.pop()

            if (map.get(popped) != char)
                return false
        }
    }
    if (stack.length > 0)
        return false
    return true
};